crates/core/tedge_actors/src/servers/message_boxes.rs

```diff
@@ -1,72 +1,65 @@
 use crate::Message;
 use crate::MessageReceiver;
-use crate::NoConfig;
+use crate::NoMessage;
+use crate::NullSender;
+use crate::ReplyToRequester;
+use crate::RequestEnvelope;
 use crate::RuntimeRequest;
 use crate::Sender;
-use crate::ServiceConsumer;
 use crate::ServiceProvider;
-use crate::SimpleMessageBox;
-use crate::SimpleMessageBoxBuilder;
+use async_trait::async_trait;
+use futures::channel::oneshot;
 use futures::StreamExt;
 use std::fmt::Debug;
 use std::ops::ControlFlow;
-use tokio::select;
 
 /// A message box for a request-response server
-pub type ServerMessageBox<Request, Response> =
-    SimpleMessageBox<(ClientId, Request), (ClientId, Response)>;
-
-/// Internal id assigned to a client actor of a server actor
-pub type ClientId = usize;
+pub type ServerMessageBox<Request, Response> = LoggingReceiver<RequestEnvelope<Request, Response>>;
 
 /// A message box for services that handles requests concurrently
 pub struct ConcurrentServerMessageBox<Request: Debug, Response> {
     /// Max concurrent requests
     max_concurrency: usize,
 
     /// Message box to interact with clients of this service
-    clients: ServerMessageBox<Request, Response>,
+    requests: ServerMessageBox<Request, Response>,
 
     /// Pending responses
-    pending_responses: futures::stream::FuturesUnordered<PendingResult<(usize, Response)>>,
+    pending_responses: futures::stream::FuturesUnordered<PendingResult>,
 }
 
-type PendingResult<R> = tokio::task::JoinHandle<R>;
+type PendingResult = tokio::task::JoinHandle<()>;
 
 impl<Request: Message, Response: Message> ConcurrentServerMessageBox<Request, Response> {
     pub(crate) fn new(
         max_concurrency: usize,
-        clients: ServerMessageBox<Request, Response>,
+        requests: ServerMessageBox<Request, Response>,
     ) -> Self {
         ConcurrentServerMessageBox {
             max_concurrency,
-            clients,
+            requests,
             pending_responses: futures::stream::FuturesUnordered::new(),
         }
     }
 
-    pub async fn recv(&mut self) -> Option<(ClientId, Request)> {
-        self.next_request().await
-    }
-
-    pub async fn send(&mut self, message: (ClientId, Response)) -> Result<(), ChannelError> {
-        self.clients.send(message).await
-    }
-
-    async fn next_request(&mut self) -> Option<(usize, Request)> {
+    pub async fn next_request(&mut self) -> Option<RequestEnvelope<Request, Response>> {
         if self.await_idle_processor().await.is_break() {
             return None;
         }
 
         loop {
             tokio::select! {
-                Some(request) = self.clients.recv() => {
+                Some(request) = self.requests.recv() => {
                     return Some(request);
                 }
                 Some(result) = self.pending_responses.next() => {
-                    self.send_result(result).await;
+                    if let Err(err) = result {
+                        log::error!("Request failed with: {err}");
```

Comment:
Not sure if that log would be very helpful as the context is missing on which request failed. But not sure if it's easy to get that request info here. I think it's better to just log the errors in the spawned task itself, which has that context, instead of ignoring it, as done now.

From what I understand only the `send` failures will reach here. And I'm assuming we're just logging it and proceeding here with the hope that the `send` failure is limited to justa single request.

Reply:
Unfortunately, I don't think we can do better.

- This error is not a error returned by a failed request. Such errors are handled inside the spawned task and forwarded to the requesting actor.
- This is not a `send` failure. These are also handled inside the spawned task (they are ignored in fact, as this just means the requester is no more expecting a response).
- This is a [`JoinError`](https://docs.rs/tokio/latest/tokio/task/struct.JoinError.html) caused by a panic while processing the request.

That said, I changed the error message to better match what is going on:

```
log::error!("Fail to run a request to completion: {err}");
```

See https://github.com/thin-edge/thin-edge.io/pull/2600/commits/606caa8a1bb5430a1ff0fd92fed460579443a9c5